ethcore/src/engines/mod.rs

```diff
@@ -370,6 +372,10 @@ pub trait Engine<M: Machine>: Sync + Send {
 		None
 	}
 
+		/// Allow hook to set extra data before closing the block (after txs are applied).  used by
+		/// clique for block signing
+		fn close_block_extra_data(&self, _parent: &M::Header) -> Option<Vec<u8>> { None }
```

Comment:
There's no need for this function. You can just do in `Engine::on_close_block` with `block.header.set_extra_data`.

Reply:
It's definitely hacky.  However, the extra data for clique contains a signature over the sealed block header so there needs to be a hook to add this signature after all header fields are finalized and transactions applied.

Reply:
I see. `on_close_block` is currently called before finalizing header fields.

However, I think we can just move [this line](https://github.com/paritytech/parity-ethereum/blob/c9e651e23b050f08c7380545876263e01b0432a6/ethcore/src/block.rs#L390) to after header fields are set. This should allow you to validate the full header in `on_close_block` directly.

I checked ethash and aura. None of them rely on "unset" header fields behavior in `on_close_block`.